### PR TITLE
Add some steps to the atom watcher

### PR DESCRIPTION
**Pull request by nono**

### CODE REVIEW

**Comment by nono** [comment]
[Ajout des steps initial diff et await write finish pour l'atom watcher](https://trello.com/c/CqNu2uUv/1303-ajout-des-steps-initial-diff-et-await-write-finish-pour-latom-watcher)

File: core/local/steps/dispatch.js [hunk 1]
```diff
@@ -54,34 +61,95 @@ actions = {
...
 
   renameddirectory: async (event) => {
-    // TODO we don't have stats
-    const src = buildDir(event.oldPath, event.stats)
-    const doc = buildDir(event.path, event.stats)
-    await target.moveFolderAsync(SIDE, doc, src)
+    let doc, old
+    try {
+      old = await fetchOldDoc(id(event.oldPath))
+      doc = buildDir(event.path, event.stats)
```
Comment (taratatach) [inline core/local/steps/dispatch.js]: Same here.

File: core/local/steps/event.js [hunk 2]
```diff
@@ -9,9 +9,9 @@ export type AtomWatcherEvent = {
   path: string,
   oldPath?: string,
   _id?: string,
-  docType?: string,
+  docType?: "file" | "directory",
```
Comment (taratatach) [inline core/local/steps/event.js]: :+1:

File: test/property/local_watcher/index.js [hunk 3]
```diff
@@ -71,8 +86,12 @@ async function step (state, op) {
     case 'mv':
       try {
         // XXX fs-extra move can enter in an infinite loop for some stupid moves
```
Comment (taratatach) [inline test/property/local_watcher/index.js]: We should remove this comment.

File: core/local/atom_watcher.js [hunk 4]
```diff
@@ -51,20 +49,25 @@ module.exports = class AtomWatcher {
     this.ignore = ignore
     this.checksumer = checksumer.init()
 
+    // Here, we build a chain of steps. Each step can be seen as an actor that
+    // communicates with the next one via a buffer. The first step is called
+    // the producer or runner: even if the chain is ready at the end of this
+    // constructor, the runner won't start pushing batches of events until it
+    // is started.
     let steps
+    // TODO add a debounce layer (a port of awaitWriteFinish of chokidar)
```
[the remark this answers is not in the document]
Comment (nono) [reply core/local/atom_watcher.js]: Yes, I forgot to remove the TODO.

File: test/property/local_watcher/index.js (hunk 3 above)
Comment (nono) [reply test/property/local_watcher/index.js]: I'd prefer to keep it. Someone that see that fs-extra used a lot in this file, and not here, may want to change it, and the mistake is hard to detect.

File: core/local/steps/await_write_finish.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      if (event.action === 'initial-scan-done' || event.kind !== 'file') {
+        batch.push(event)
+        continue
+      }
+
+      let item /*: ?PendingItem */ = pending.get(event._id)
+      if (['created', 'modified'].includes(event.action)) {
+        if (item) {
+          item.event = event
+          clearTimeout(item.timeout)
```
[the remark this answers is not in the document]
Comment (taratatach) [reply core/local/steps/await_write_finish.js]: It's not mandatory. But the current code requires some good knowledge of the event loop. I'll review the links I have about it :D.

File: test/property/local_watcher/index.js (hunk 3 above)
Comment (taratatach) [reply test/property/local_watcher/index.js]: OK. I'm fine with that.

File: core/local/steps/add_checksum.js [hunk 6]
```diff
@@ -13,7 +18,8 @@ module.exports = function (buffer /*: Buffer */, opts /*: { syncPath: string , c
     const batch = []
     for (const event of events) {
       try {
-        if (['created', 'modified', 'scan'].includes(event.action) && event.docType === 'file') {
+        if (['created', 'modified', 'scan', 'renamed'].includes(event.action) && event.docType === 'file') {
```
[the remark this answers is not in the document]
Comment (taratatach) [reply core/local/steps/add_checksum.js]: OK. This is probably something we should test if we can :)

**Pull request merged**